core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/AggregateFunction.java

```diff
@@ -16,27 +16,46 @@
 
 package io.cdap.plugin.batch.aggregator.function;
 
+import com.google.common.collect.Table;
 import io.cdap.cdap.api.data.format.StructuredRecord;
 import io.cdap.cdap.api.data.schema.Schema;
 
+import javax.annotation.Nullable;
+
 /**
- * Performs an aggregation. For each group that needs an aggregate to be calculated, the {@link #beginFunction()}
- * method is called first. After that, one or more calls to {@link #operateOn(StructuredRecord)} are made, one call for
- * each value in the group. Finally, {@link #getAggregate()} is called to retrieve the aggregate.
- *
- * todo: convert this to a plugin
+ * Performs an aggregation.
  *
  * @param <T> type of aggregate value
  */
-public interface AggregateFunction<T> extends RecordFunctionLifecycle {
+public interface AggregateFunction<T> {
+
+  /**
+   * Merge the given record to the aggregated collection. This function is guaranteed to get called at least once
+   * before {@link #mergeAggregates(Table, Table)} is called.
+   *
+   * @param agg the aggregated collection, the merged result should be added to it
+   * @param record the record to merge
+   */
+  void mergeValue(Table<String, String, Object> agg, StructuredRecord record);
```

Comment:
After reading some more it seems like it's used to carry intermediate data? Using Table as the intermediate data makes things harder to understand and maintain because it's not clear what is in the Table. 

It seems like it would be better for each individual function to be able to decide what reduced data it needs to keep track of, using an actual object instead of a Table or a Map with special keys.